### PR TITLE
Breaking changes 3.0

### PR DESCRIPTION
**Pull request by JarnoWesthof**
Changelist:
* Added DataStorageHelper 
* Added RoomStateHelper 
* Added ConnectionInfoHelper
* Changed datatype of location and itemIds to long
* UpdateTags dropped in favor of UpdateConnectionOptions
* UpdateConnectionOptions now requires the itemsHandlingFlags to be specified
* Removed MissingChecks & LocationsChecked from LoginSuccessful as they are now available on the LocationCheckHelper
* ItemsHandling on ConnectPacket is now mandatory
* Changed all lists inside data packages to arrays since they represent an immutable state
* Changed datatype of location and itemIds to long
* Arrays in packages that are not provided by the server are now null rather than empty
* Tags and ItemHandlingFlags moved to ConnectionInfo instead of Session

### CODE REVIEW

**Comment by JarnoWesthof** [comment]
Still need testing now everything is merged together

**Comment by Ijwu** [comment]
Are we okay to merge this? Have you had the chance to play with it? Sorry I haven't been around much lately.

**Comment by JarnoWesthof** [comment]
Merged in the DataStorage + updated the readme to include latest changes

File: README.md [hunk 1]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+session.DataStorage["Min"] = (session.DataStorage["Min"] + 40) >> 100; //Add 40 to `Min`, then Set `Min` to 100 if `Min` is bigger than 100
+session.DataStorage["C"] = ((session.DataStorage["C"] - 6) + Bitwise.RightShift(1)) ^ 3; //Subtract 6 from `C`, then multiply `C` by 2 using bitshifting, then take `C` to the power of 3
+
+//Update callbacks
+//EnergyLink deplete pattern, subtract 50, then set value to 0 if its lower than 0
+session.DataStorage["EnergyLink"] = ((session.DataStorage["EnergyLink"] - 50) << 0) + Callback.Add((old, new) => {
+    var actualDepleted = (float)new - (float)old; //calculate the actual change, might differ if there was less than 50 left on the server
+});
+
+//Keepking track of changes
```
Comment (Ijwu) [inline README.md:167]:
```suggestion
//Keeping track of changes
```

File: README.md [hunk 2]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+
+//Storing/Updating
+session.DataStorage[Scope.Slot, "SetPersonal"] = 20; //Set `SetPersonal` to 20, in scope of the current connected user\slot
+session.DataStorage[Scope.Global, "SetGlobal"] = 30; //Set `SetGlobal` to 30, in global scope shared among all players (the default scope is global)
+session.DataStorage["Add"] += 50; //Add 50 to the current value of `Add`
+session.DataStorage["Divide"] /= 2; //Divide current value of `Divide` in half
+session.DataStorage["Max"] <<= 80; //Set `Max` to 80 if the stored value is lower than 80
+session.DataStorage["Dictionary"] = JObject.FromObject(new Dictionary<string, int>()); //Set `Dictionary` to a Dictionary
+session.DataStorage["SetObject"] = JObject.FromObject(new SomeClassOrStruct()); //Set `SetObject` to a custom object
+session.DataStorage["BitShiftLeft"] += Bitwise.LeftShift(1); //Bitshift current value of  `BitShiftLeft` to left by 1
```
Comment (Ijwu) [inline README.md:151]:
```suggestion
session.DataStorage["BitShiftLeft"] += Bitwise.LeftShift(1); //Bitshift current value of `BitShiftLeft` to left by 1
```

File: Archipelago.MultiClient.Net/BounceFeatures/DeathLink/DeathLinkService.cs [hunk 3]
```diff
@@ -14,30 +14,44 @@ public class DeathLinkService
...
+        private void OnPacketReceived(ArchipelagoPacketBase packet)
         {
-            if (packet.PacketType != ArchipelagoPacketType.Bounced)
+            switch (packet)
             {
-                return;
-            }
+                case ConnectedPacket connected:
+                    slot = connected.Slot;
+                    logger["DeathLinkReceived" + slot].Initialize(new string[0]);
```
Comment (Ijwu) [inline Archipelago.MultiClient.Net/BounceFeatures/DeathLink/DeathLinkService.cs]: Should you add slot to the key here manually or just use a Slot scope?

Comment (JarnoWesthof) [reply Archipelago.MultiClient.Net/BounceFeatures/DeathLink/DeathLinkService.cs]: Was already changed to Scope.Slot in a later commit

**Comment by Ijwu** [comment]
This is a great load of work and I'm super happy with how this is turning out. Thanks for your amazing contributions.

**Pull request merged**